### PR TITLE
fix coordinator informer panic with ServiceCIDR in low k8s version

### LINKED ISSUE #3263: coordinator informer panic
### Spiderpool Version

main

### Bug Type

Coordinator

### Main CNI

/

### What happened?

kubernetes version: `v1.27.0`
spiderpool main branch

```text
{"level":"INFO","ts":"2024-02-29T09:15:07.835Z","logger":"Subnet-Informer","caller":"subnetmanager/subnet_informer.go:294","msg":"Succeed to SYNC","SubnetName":"default-v4-subnet","Operation":"PROCESS"}
{"level":"INFO","ts":"2024-02-29T09:15:07.835Z","logger":"Subnet-Informer","caller":"subnetmanager/subnet_informer.go:294","msg":"Succeed to SYNC","SubnetName":"default-v6-subnet","Operation":"PROCESS"}
I0229 09:15:07.838840       1 shared_informer.go:318] Caches are synced for SpiderCoordinator
{"level":"INFO","ts":"2024-02-29T09:15:07.848Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:463","msg":"Trying to fetch the ClusterCIDR from kube-system/kubeadm-config","Event Key":"SpiderCoordinator/default","Operation":"PROCESS"}
{"level":"INFO","ts":"2024-02-29T09:15:07.848Z","logger":"Coordinator-Informer","caller":"coordinatormanager/coordinator_informer.go:484","msg":"Detect podCIDRType is: none, try to update podCIDR","Event Key":"SpiderCoordinator/default","Operation":"PROCESS"}
E0229 09:15:07.849097       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 481 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x209bbe0?, 0x399a8a0})
	/src/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00121e110?})
	/src/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x75
panic({0x209bbe0, 0x399a8a0})
	/usr/local/go/src/runtime/panic.go:884 +0x213
github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).updateServiceCIDR(0xc0008cc0f0, 0xc00059ac40, 0xc0008688c0)
	/src/pkg/coordinatormanager/coordinator_informer.go:709 +0x5d
github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).updatePodAndServerCIDR(0xc0008cc0f0, {0x27b0838, 0xc001330150}, 0xc00059ac40, 0xc00121a000?)
	/src/pkg/coordinatormanager/coordinator_informer.go:507 +0xb8b
github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).syncHandler(0xc0008cc0f0, {0x27b0838, 0xc001330150})
	/src/pkg/coordinatormanager/coordinator_informer.go:433 +0xe6
github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).processNextWorkItem(0xc0008cc0f0, {0x27b0838, 0xc0008200c0})
	/src/pkg/coordinatormanager/coordinator_informer.go:414 +0x30c
github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).runWorker(...)
	/src/pkg/coordinatormanager/coordinator_informer.go:398
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x1f259a0?, {0x278ffc0, 0xc0013300f0}, 0x1, 0xc00054e540)
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x1?, 0x3b9aca00, 0x0, 0x90?, 0x200000000?)
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x27b0838, 0xc0008200c0}, 0xc0006f9e50, 0xc000779080?, 0x0?, 0x0?)
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x27b0838?, 0xc0008200c0?}, 0xc000611768?, 0xc000611778?)
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:170 +0x2b
created by github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).run
	/src/pkg/coordinatormanager/coordinator_informer.go:389 +0x285
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1db933d]

goroutine 481 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00121e110?})
	/src/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xd7
panic({0x209bbe0, 0x399a8a0})
	/usr/local/go/src/runtime/panic.go:884 +0x213
github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).updateServiceCIDR(0xc0008cc0f0, 0xc00059ac40, 0xc0008688c0)
	/src/pkg/coordinatormanager/coordinator_informer.go:709 +0x5d
github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).updatePodAndServerCIDR(0xc0008cc0f0, {0x27b0838, 0xc001330150}, 0xc00059ac40, 0xc00121a000?)
	/src/pkg/coordinatormanager/coordinator_informer.go:507 +0xb8b
github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).syncHandler(0xc0008cc0f0, {0x27b0838, 0xc001330150})
	/src/pkg/coordinatormanager/coordinator_informer.go:433 +0xe6
github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).processNextWorkItem(0xc0008cc0f0, {0x27b0838, 0xc0008200c0})
	/src/pkg/coordinatormanager/coordinator_informer.go:414 +0x30c
github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).runWorker(...)
	/src/pkg/coordinatormanager/coordinator_informer.go:398
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x1f259a0?, {0x278ffc0, 0xc0013300f0}, 0x1, 0xc00054e540)
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x1?, 0x3b9aca00, 0x0, 0x90?, 0x200000000?)
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x27b0838, 0xc0008200c0}, 0xc0006f9e50, 0xc000779080?, 0x0?, 0x0?)
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x27b0838?, 0xc0008200c0?}, 0xc000611768?, 0xc000611778?)
	/src/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:170 +0x2b
created by github.com/spidernet-io/spiderpool/pkg/coordinatormanager.(*CoordinatorController).run
	/src/pkg/coordinatormanager/coordinator_informer.go:389 +0x285
```

### What did you expect to happen?

/

### How to reproduce it (as minimally and precisely as possible)

/

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Icarus9913**
Signed-off-by: Icarus9913 [icaruswu66@qq.com](mailto:icaruswu66@qq.com)

**What this PR does / why we need it**:
fix bug

**Which issue(s) this PR fixes**:
close #3263

### CODE REVIEW

File: pkg/coordinatormanager/coordinator_informer.go [hunk 1]
```diff
@@ -235,19 +240,28 @@ func (cc *CoordinatorController) addEventHandlers(
...
-	if err = cc.addServiceCIDRHandler(serviceCIDRInformer.Informer()); err != nil {
-		return err
+	if cc.ServiceCIDR.IsAvailable {
+		InformerLogger.Debug("the ServiceCIDR is available in your cluster, Start the serviceCIDR informer")
+		if err = cc.addServiceCIDRHandler(serviceCIDRInformer.Informer()); err != nil {
+			return err
+		}
+	} else {
+		InformerLogger.Warn("ServiceCIDR feature is unavailable in your cluster, Don't start the serviceCIDR informer")
 	}
```
Comment (Icarus9913) [inline pkg/coordinatormanager/coordinator_informer.go:263]: Just use the kubernetes API resource to check whether the ServiceCIDR resource exists in the current cluster.
Reference previous PR: https://github.com/spidernet-io/spiderpool/pull/3185

**Review by weizhoublue — APPROVED** [review]

File: pkg/coordinatormanager/coordinator_informer.go [hunk 2]
```diff
@@ -235,19 +240,29 @@ func (cc *CoordinatorController) addEventHandlers(
...
 
+	// check the current k8s whether registered the ServiceCIDR resource
 	InformerLogger.Debug("Checking if the ServiceCIDR is available in your cluster")
-	var serviceCIDR networkingv1alpha1.ServiceCIDRList
-	err = cc.APIReader.List(context.TODO(), &serviceCIDR)
-	if err != nil {
-		InformerLogger.Warn("ServiceCIDR feature is unavailable in your cluster, Don't start the serviceCIDR informer")
-		return nil
+	resourceList, err := cc.K8sClient.DiscoveryClient.ServerResourcesForGroupVersion(networkingv1alpha1.SchemeGroupVersion.String())
+	if nil != err {
```
Comment (cyclinder) [inline pkg/coordinatormanager/coordinator_informer.go:246]:
```suggestion
	if err != nil {
```

File: pkg/coordinatormanager/coordinator_informer.go [hunk 3]
```diff
@@ -369,8 +383,8 @@ func (cc *CoordinatorController) run(ctx context.Context, workers int) error {
 	defer cc.Workqueue.ShutDown()
 
 	additionalCacheSync := []cache.InformerSynced{cc.CoordinatorSynced, cc.ConfigmapSynced}
-	if cc.ServiceCIDRSynced != nil {
-		additionalCacheSync = append(additionalCacheSync, cc.ServiceCIDRSynced)
+	if cc.ServiceCIDR.IsAvailable {
```
Comment (cyclinder) [inline pkg/coordinatormanager/coordinator_informer.go:386]: 我感觉 不需要 ServiceCIDR.IsAvailable 这个字段 ，cc.ServiceCIDRLister = serviceCIDRInformer.Lister() 可以提前赋值，而 cc.ServiceCIDRSynced = serviceCIDRInformer.HasSynced 在检查到 serviceCDIR 资源后复制。这里还是通过 if cc.ServiceCIDRSynced != nil 决定是否加入 cacheSync,

**Comment by Icarus9913** [comment]
https://github.com/spidernet-io/spiderpool/actions/runs/8105788968

Already tested in kubernetes Matrix

File: pkg/coordinatormanager/coordinator_informer.go (hunk 3 above)
Comment (Icarus9913) [reply pkg/coordinatormanager/coordinator_informer.go:386]: 对某些函数一直判断是否为空，增加代码理解成本，导致每次使用之前，还得理清前面的逻辑到底是否有给当前函数变量赋值

File: pkg/coordinatormanager/coordinator_informer.go (hunk 2 above)
Comment (Icarus9913) [reply pkg/coordinatormanager/coordinator_informer.go:246]: 一样的，代码习惯而已。不存在对错。
https://stackoverflow.com/questions/3021195/which-is-more-effective-if-null-variable-or-if-variable-null

Comment (cyclinder) [reply pkg/coordinatormanager/coordinator_informer.go:246]: 不说对错，只是建议 ( 目前 cncf 项目还没看到过这种写法

**Comment by cyclinder** [comment]
I will close it, due to dup with https://github.com/spidernet-io/spiderpool/pull/3269

**Pull request closed without merging**